### PR TITLE
Typescript: make parsing faster

### PR DESCRIPTION
**Pull request by james-rms**
### Changelog
- Typescript: improved record parsing time.

### Docs

None.

### Description

I was trying to port some of the improvements from https://github.com/foxglove/mcap/pull/1168 to typescript, and noticed that this change to `parse.ts` improves reading speed significantly with no additional API changes. 

This change switches from `new Uint8Array(fromBuffer.slice(start, end)` to `new Uint8Array(fromBuffer, start, length).slice()`. These both have the same effect of producing a new Uint8Array with a copy of some part of `fromBuffer`. 

I also rewrote the typescript benchmark utility to remove the dependency on `benny`, so that I could gather memory statistics as part of the benchmark.

#### Before:

```
(python-sNIFi2pF) j@192-168-1-105 benchmarks % yarn bench --suite reader
Running 'reader' suite
McapStreamReader
        1.29±0.02 op/s  Heap Used: 259.38±4.95 MB/op    Heap Total: 288.73±3.47 MB/op   ArrayBuffers: 214.31±1.60 MB/op
McapIndexedReader
        0.95±0.01 op/s  Heap Used: 248.20±21.02 MB/op   Heap Total: 281.67±20.46 MB/op  ArrayBuffers: 98.33±5.42 MB/op
McapIndexedReader_reverse
        1.00±0.02 op/s  Heap Used: 260.77±17.31 MB/op   Heap Total: 295.44±17.31 MB/op  ArrayBuffers: 102.49±4.62 MB/op
```
#### After

```
(python-sNIFi2pF) j@192-168-1-105 benchmarks % yarn bench --suite reader
Running 'reader' suite
McapStreamReader
        3.04±0.02 op/s  Heap Used: 261.58±8.20 MB/op    Heap Total: 289.82±6.71 MB/op   ArrayBuffers: 214.18±1.60 MB/op
McapIndexedReader
        1.83±0.01 op/s  Heap Used: 281.22±4.06 MB/op    Heap Total: 317.18±2.63 MB/op   ArrayBuffers: 107.06±0.00 MB/op
McapIndexedReader_reverse
        1.86±0.00 op/s  Heap Used: 278.37±2.01 MB/op    Heap Total: 313.65±0.93 MB/op   ArrayBuffers: 107.06±0.00 MB/op
```

### CODE REVIEW

File: typescript/benchmarks/package.json [hunk 1]
```diff
@@ -17,8 +17,8 @@
     "typecheck": "tsc -p tsconfig.json --noEmit",
     "lint:ci": "eslint --report-unused-disable-directives .",
     "lint": "eslint --report-unused-disable-directives --fix .",
-    "bench": "ts-node --files --project tsconfig.cjs.json index.ts",
-    "bench:debug": "NODE_OPTIONS='--inspect-brk' ts-node --files --project tsconfig.cjs.json index.ts"
+    "bench": "TS_NODE_PROJECT=tsconfig.cjs.json node --expose-gc -r 'ts-node/register' index.ts",
```
Comment (james-rms) [inline typescript/benchmarks/package.json:20]:
```suggestion
    "bench": "TS_NODE_FILES=true TS_NODE_PROJECT=tsconfig.cjs.json node --expose-gc -r 'ts-node/register' index.ts",
```

File: typescript/benchmarks/package.json [hunk 2]
```diff
@@ -17,8 +17,8 @@
     "typecheck": "tsc -p tsconfig.json --noEmit",
     "lint:ci": "eslint --report-unused-disable-directives .",
     "lint": "eslint --report-unused-disable-directives --fix .",
-    "bench": "ts-node --files --project tsconfig.cjs.json index.ts",
-    "bench:debug": "NODE_OPTIONS='--inspect-brk' ts-node --files --project tsconfig.cjs.json index.ts"
+    "bench": "TS_NODE_FILES=true TS_NODE_PROJECT=tsconfig.cjs.json node --expose-gc -r 'ts-node/register' index.ts",
+    "bench:debug": "TS_NODE_PROJECT=tsconfig.cjs.json node --inspect-brk --expose-gc -r 'ts-node/register' index.ts"
```
Comment (james-rms) [inline typescript/benchmarks/package.json:21]:
```suggestion
    "bench:debug": "TS_NODE_FILES=true TS_NODE_PROJECT=tsconfig.cjs.json node --inspect-brk --expose-gc -r 'ts-node/register' index.ts"
```

File: typescript/core/src/parse.ts [hunk 3]
```diff
@@ -62,10 +62,15 @@ export function parseRecord({
...
   if (!isKnownOpcode(opcode)) {
+    const data = new Uint8Array(
+      view.buffer,
+      view.byteOffset + headerReader.offset,
+      recordLengthNum,
+    ).slice();
     const record: TypedMcapRecord = {
       type: "Unknown",
       opcode,
-      data: new Uint8Array(view.buffer, view.byteOffset + headerReader.offset, recordLengthNum),
```
Comment (defunctzombie) [inline typescript/core/src/parse.ts:68]: This did not previously copy the buffer - do we want to be doing that copy now?

File: typescript/core/src/parse.ts [hunk 4]
```diff
@@ -107,10 +112,9 @@ export function parseRecord({
         throw new Error(`Schema data length ${dataLen} exceeds bounds of record`);
       }
       const data = new Uint8Array(
-        recordView.buffer.slice(
```
Comment (defunctzombie) [inline typescript/core/src/parse.ts:110]: We've lost this copy - intentional?

File: typescript/core/src/parse.ts (hunk 3 above)
Comment (james-rms) [reply typescript/core/src/parse.ts:68]: I updated this to restore the original behavior for comparison reasons, but I believe this should be a copy.

File: typescript/core/src/parse.ts (hunk 4 above)
Comment (james-rms) [reply typescript/core/src/parse.ts:110]: I restored this, benchmark data updated.

**Review by defunctzombie — APPROVED** [review]

**Pull request merged**